### PR TITLE
Backfilling performance

### PR DESCRIPTION
**Pull request by eesmerdag**
https://www.notion.so/figmentnetworks/Find-a-best-way-to-speed-up-the-process-of-backfilling-data-354c1ef5a2c84b708d281159d0d29393

### CODE REVIEW

File: indexer/pipeline.go [hunk 1]
```diff
@@ -196,65 +195,85 @@ func (p *indexingPipeline) Backfill(ctx context.Context, backfillCfg BackfillCon
...
 	pipelineOptionsCreator := &pipelineOptionsCreator{
 		configParser:      p.configParser,
 		desiredVersionIds: versionIds,
 	}
-	pipelineOptions, err := pipelineOptionsCreator.parse()
+	return pipelineOptionsCreator.parse()
+}
+
+// generates report if not exists
+func (p *indexingPipeline) generateReportCreatureAndSource(indexVersion int64, isParallel bool) (*reportCreator, *backfillSource, error) {
```
Comment (pacmessica) [inline indexer/pipeline.go:244]: Creator, not creature :)

File: indexer/source_backfill.go [hunk 2]
```diff
@@ -60,43 +59,37 @@ func (s *backfillSource) Err() error {
...
+	if err := s.setEndSyncsOfSessionsAndEras(); err != nil {
 		return err
 	}
+	s.setCurrentHeight()
+	s.setEndHeight()
 	return nil
 }
 
-func (s *backfillSource) setStartHeight() error {
-	syncable, err := s.db.Syncables.FindFirstByDifferentIndexVersion(s.currentIndexVersion)
```
Comment (pacmessica) [inline indexer/source_backfill.go:78]: backfill should start up from wherever it last stopped, that's what `FindFirstByDifferentIndexVersion` is doing. If we run backfill for the first time, then it'll start from height 0, but if we stop the backfill process and then restart it shouldn't start from 0 again.

**Review by pacmessica — COMMENTED** [review]
It looks like you made changes to only run backfill for end of session/era heights. 

I guess the ticket scope wasn't clear enough. There should be some kind of flag if we only want to run backfill fetching data from session/era heights. It should still be possible to run backfill for every height like it was before  :)

File: indexer/source_backfill.go (hunk 2 above)
Comment (pacmessica) [reply indexer/source_backfill.go:78]: When we want to run backfill, what we do is add another version to the indexer_config.json file. Whenever the indexer runs (eg worker or backfill), it tracks which heights its collected data for and which version its used (this is saved in reports table).

https://github.com/figment-networks/polkadothub-indexer/blob/master/indexer_config.json

**Comment by eesmerdag** [comment]
> It looks like you made changes to only run backfill for end of session/era heights.
> 
> I guess the ticket scope wasn't clear enough. There should be some kind of flag if we only want to run backfill fetching data from session/era heights. It should still be possible to run backfill for every height like it was before :)

So lets say;
1. 100th height is indexed and suppose these heights are the end of validators (lets say 3) 30, 60, 90.
2. And we decided to fetch a new column for our database from nodes while the latest height is 100.
3. Then should not i process only the heights 30, 60 and 90 since only related data is onn these heights?

File: indexer/pipeline.go (hunk 1 above)
Comment (eesmerdag) [reply indexer/pipeline.go:244]: :alien:
:alien: :alien:
:alien: :alien: :alien:
:alien: :alien: :alien: :alien:
:alien: :alien: :alien: :alien: :alien:
:alien: :alien: :alien: :alien:
:alien: :alien: :alien:
:alien: :alien:
:alien:

**Review by silentlight — COMMENTED** [review]
@pacmessica put it right. We probably want to be able to pass in some flags to make it all work. Maybe let's chat about it today all 3 of us?

**Comment by pacmessica** [comment]
> > It looks like you made changes to only run backfill for end of session/era heights.
> > I guess the ticket scope wasn't clear enough. There should be some kind of flag if we only want to run backfill fetching data from session/era heights. It should still be possible to run backfill for every height like it was before :)
> 
> So lets say;
> 
> 1. 100th height is indexed and suppose these heights are the end of validators (lets say 3) 30, 60, 90.
> 2. And we decided to fetch a new column for our database from nodes while the latest height is 100.
> 3. Then should not i process only the heights 30, 60 and 90 since only related data is onn these heights?

if you added a column to the validators table, then yes, you should only process data on heights 30,60,90.

But what if we added a column to the events table? Events happen at every height, so we need to run backfill from heights 1-90.

**Pull request closed without merging**